core/src/main/java/io/grpc/ServerInterceptors.java

```diff
@@ -179,101 +173,96 @@ public InputStream parse(final InputStream stream) {
   public static <T> ServerServiceDefinition useMarshalledMessages(
       final ServerServiceDefinition serviceDef,
       final MethodDescriptor.Marshaller<T> marshaller) {
-    final ServerServiceDefinition.Builder serviceBuilder = ServerServiceDefinition
-        .builder(serviceDef.getName());
-    for (final ServerMethodDefinition<?, ?> definition : serviceDef.getMethods()) {
-      final MethodDescriptor<?, ?> originalMethodDescriptor = definition.getMethodDescriptor();
-      final MethodDescriptor<T, T> wrappedMethodDescriptor = MethodDescriptor
-          .create(originalMethodDescriptor.getType(),
-              originalMethodDescriptor.getFullMethodName(),
+    List<MethodDescriptor> wrappedMethods = new ArrayList<MethodDescriptor>(
+        serviceDef.getDescriptor().getMethods().size());
+    for (final MethodDescriptor<?, ?> originalMethod : serviceDef.getDescriptor().getMethods()) {
+      MethodDescriptor<T, T> wrappedMethodDescriptor = MethodDescriptor
+          .create(originalMethod.getType(),
+              originalMethod.getFullMethodName(),
+              marshaller,
               marshaller,
-              marshaller);
-      serviceBuilder.addMethod(wrapMethod(definition, wrappedMethodDescriptor));
+              originalMethod.getServiceIndex());
+      wrappedMethods.add(wrappedMethodDescriptor);
     }
-    return serviceBuilder.build();
+    return new ServerServiceDefinition(
+        new ServiceDescriptor(serviceDef.getDescriptor().getName(), wrappedMethods),
+        wrapHandler(serviceDef.getDescriptor(), serviceDef.getCallHandler()));
   }
 
-  private static <ReqT, RespT> void wrapAndAddMethod(
-      ServerServiceDefinition.Builder serviceDefBuilder, ServerMethodDefinition<ReqT, RespT> method,
+  private static ServerCallHandler wrapServerCallHandler(
+      ServerCallHandler toWrap,
       List<? extends ServerInterceptor> interceptors) {
-    ServerCallHandler<ReqT, RespT> callHandler = method.getServerCallHandler();
     for (ServerInterceptor interceptor : interceptors) {
-      callHandler = InterceptCallHandler.create(interceptor, callHandler);
+      toWrap = InterceptCallHandler.create(interceptor, toWrap);
     }
-    serviceDefBuilder.addMethod(method.withServerCallHandler(callHandler));
+    return toWrap;
   }
 
-  private static class InterceptCallHandler<ReqT, RespT> implements ServerCallHandler<ReqT, RespT> {
-    public static <ReqT, RespT> InterceptCallHandler<ReqT, RespT> create(
-        ServerInterceptor interceptor, ServerCallHandler<ReqT, RespT> callHandler) {
-      return new InterceptCallHandler<ReqT, RespT>(interceptor, callHandler);
+  private static class InterceptCallHandler implements ServerCallHandler {
+
+    public static <ReqT, RespT> InterceptCallHandler create(
```

Comment:
`<ReqT, RespT>` here looks unused.

The original reason for the create method was to simplify dealing with the generics. Feel free to remove it and use the constructor directly.